// mathjax-angular/esm2022/directive/mathjax.module.mjs
import { NgModule } from '@angular/core';
import { MathjaxDirective } from './mathjax.directive';
import { MathjaxDefaultConfig, mathjax_url, RootMathjaxConfig, } from '../models';
import * as i0 from "@angular/core";
import * as i1 from "../models";
export class MathjaxModule {
    constructor(moduleConfig) {
        this.moduleConfig = moduleConfig;
        //
        this.addConfigToDocument();
        //
        this.addMatjaxToDocument();
    }
    addConfigToDocument() {
        const tagId = 'mathjax-config-script';
        const isScript = document.getElementById(tagId);
        if (isScript)
            return;
        //
        const providConfig = {
            ...MathjaxDefaultConfig,
            ...(this.moduleConfig?.config ?? {}),
        };
        const script = document.createElement('script');
        script.id = tagId;
        script.type = 'text/javascript';
        script.text = `
      MathJax = ${JSON.stringify(providConfig)};
      MathJax.isReady = false;
      MathJax.promise = new Promise(function (resolve, reject) {
        MathJax.startup = {
            ready() {
              MathJax.isReady = true;
              MathJax.startup.defaultReady();
              MathJax.startup.promise.then(() => resolve());
            }
        };
      });
    `;
        document.getElementsByTagName('head')[0].appendChild(script);
    }
    addMatjaxToDocument() {
        const tagId = 'mathjax-script';
        const isScript = document.getElementById(tagId);
        if (isScript)
            return;
        //
        const script = document.createElement('script');
        script.id = tagId;
        script.type = 'text/javascript';
        script.src = this.moduleConfig?.src ?? mathjax_url;
        script.async = true;
        document.getElementsByTagName('head')[0].appendChild(script);
    }
    static forRoot(config) {
        return {
            ngModule: MathjaxModule,
            providers: [{ provide: RootMathjaxConfig, useValue: config ?? {} }],
        };
    }
    static forChild() {
        return {
            ngModule: MathjaxModule,
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: MathjaxModule, deps: [{ token: i1.RootMathjaxConfig }], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.1.3", ngImport: i0, type: MathjaxModule, declarations: [MathjaxDirective], exports: [MathjaxDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: MathjaxModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: MathjaxModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MathjaxDirective],
                    exports: [MathjaxDirective],
                }]
        }], ctorParameters: () => [{ type: i1.RootMathjaxConfig }] });
