// mathjax-angular/esm2022/directive/mathjax.directive.mjs
import { Directive, Input, } from '@angular/core';
import { fixMathjaxBugs, getMathjaxContent, isMathjax } from '../utils';
import * as i0 from "@angular/core";
export class MathjaxDirective {
    //
    set mathjax(val) {
        this.mathJaxExpressions = val;
    }
    //
    constructor(el) {
        this.el = el;
        //
        this.element = el.nativeElement;
    }
    ngOnChanges(changes) {
        const expressions = changes.mathjax;
        if (!expressions ||
            expressions.currentValue === expressions.previousValue) {
            return;
        }
        //
        const value = getMathjaxContent(expressions.currentValue) + '';
        //
        if (isMathjax(value)) {
            const filteredVal = fixMathjaxBugs(value);
            this.typeset(() => {
                this.element.innerHTML = `<div class='jax-process'>${filteredVal}</div>`;
            });
        }
        else {
            this.element.innerHTML = value;
        }
    }
    typeset(code) {
        if (!MathJax?.isReady) {
            this.element.innerHTML = `<div></div>`;
            return MathJax.promise.then(() => this.typeset(code));
        }
        else {
            MathJax.startup.promise = MathJax.startup.promise
                .then(() => MathJax.typesetPromise(code()))
                .catch((err) => console.log('Typeset failed: ' + err.message));
            return MathJax.startup.promise;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: MathjaxDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.1.3", type: MathjaxDirective, selector: "mathjax,[mathjax]", inputs: { mathjax: "mathjax" }, usesOnChanges: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: MathjaxDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'mathjax,[mathjax]',
                }]
        }], ctorParameters: () => [{ type: i0.ElementRef }], propDecorators: { mathjax: [{
                type: Input,
                args: ['mathjax']
            }] } });
